### PR TITLE
Add `azure_rm_afdendpoint` module to support Azure Front Door Standard and Premium

### PR DESCRIPTION
**Pull request by jartoo**
##### SUMMARY

This module manages Frontdoor Endpoints for the Standard and Premium services.  I propose building this as a separate module in lieu of complicating the azure_cdn_profile module.

Begins to work on https://github.com/ansible-collections/azure/issues/1041 which will require additional pull requests and new modules. This does not complete this issue, however.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

Module: azure_rm_afdendpoint

##### ADDITIONAL INFORMATION

Provides the ability to manage Endpoints, per the Python SDK here: https://learn.microsoft.com/en-us/python/api/azure-mgmt-cdn/azure.mgmt.cdn.operations.afdendpointsoperations?view=azure-python

```paste below

```

### CODE REVIEW

File: tests/integration/targets/azure_rm_afdendpoint/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    that:
+      - not output.changed
+
+- name: Update EndPoint (replace tags)
+  azure_rm_afdendpoint:
+    name: "{{ endpoint }}"
+    profile_name: "{{ profile }}"
+    resource_group: "{{ resource_group }}"
+    enabled_state: "Disabled"
+    append_tags: False
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_afdendpoint/tasks/main.yml]: Truthy value should be one of
```suggestion
    append_tags: false
```

File: tests/integration/targets/azure_rm_afdendpoint/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    that:
+      - not output.changed
+
+- name: Delete Frontdoor Profile
+  azure_rm_cdnprofile:
+    name: "{{ profile }}"
+    location: "Global"
+    resource_group: "{{ resource_group }}"
+    state: "absent"
+
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_afdendpoint/tasks/main.yml]: Too many blank lines
```suggestion
```

File: plugins/modules/azure_rm_afdendpoint_info.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+DOCUMENTATION = '''
+---
+module: azure_rm_afdendpoint_info
+
+version_added: ""
+
+short_description: Get Azure Front Door Endpoint facts to be used with Standard or Premium Frontdoor Service
+
+description:
+    - Get facts for a specific Azure Front Door (AFD) Endpoint or all AFD Endpoints.  This differs from the Front Door classic service and only is intended to be used by the Standard or Premium service offering.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint_info.py]: line too long (211 > 160 characters)

File: plugins/modules/azure_rm_afdendpoint_info.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    profile_name: myCDNProfile
+    name: myEndpoint1
+'''
+
+RETURN = '''
+afdendpoints:
+    description: List of AFD Endpoints.
+    returned: always
+    type: complex
+    contains:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint_info.py]: trailing whitespace
```suggestion
    contains:
```

File: plugins/modules/azure_rm_afdendpoint_info.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            facts_module=True
+        )
+
+    def exec_module(self, **kwargs):
+
+        for key in self.module_args:
+            setattr(self, key, kwargs[key])
+
+        self.endpoint_client = self.get_mgmt_svc_client(CdnManagementClient,
+                                                   base_url=self._cloud_environment.endpoints.resource_manager,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint_info.py]: continuation line under-indented for visual indent

File: plugins/modules/azure_rm_afdendpoint_info.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                                                   api_version='2023-05-01')
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint_info.py]: continuation line under-indented for visual indent
```suggestion
                                                   api_version='2023-05-01')
```

File: plugins/modules/azure_rm_afdendpoint_info.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_afdendpoint_info
+
+version_added: ""
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint_info.py]: Missing version
```suggestion
version_added: ""
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+# TODO: Name check the URL
+# 
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: azure_rm_afdendpoint
+
+version_added: ""
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: Missing version
```suggestion
version_added: ""
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+short_description: Manage an Azure Front Door Endpoint to be used with Standard or Premium Frontdoor
+
+description:
+    - Create, update and delete an Azure Front Door (AFD) Endpoint to be used by a Front Door Service Profile created using azure_rm_cdnprofile.  This differs from the Front Door classic service and only is intended to be used by the Standard or Premium service offering.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: line too long

File: plugins/modules/azure_rm_afdendpoint.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+
+options:
+    auto_generated_domain_name_label_scope:
+        description:
+            - Indicates the endpoint name reuse scope. Known values are: "TenantReuse", "SubscriptionReuse", "ResourceGroupReuse", and "NoReuse". Cannot be used to update an existing Endpoint at this time.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: line too long

File: plugins/modules/azure_rm_afdendpoint.py [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    resource_group: myResourceGroup
+    state: absent
+'''
+RETURN = '''
+id:
+    description:
+        - ID of the AFD Endpoint.
+    returned: always
+    type: str
+    sample: "id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourcegroups/myResourceGroup/providers/Microsoft.Cdn/profiles/myProfile/endpoints/myEndpoint"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: line too long

File: plugins/modules/azure_rm_afdendpoint.py [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+    from azure.mgmt.cdn.models import AFDEndpoint, AFDEndpointUpdateParameters
+    from azure.mgmt.cdn import CdnManagementClient
+except ImportError as ec:
+    # This is handled in azure_rm_common
+    pass
+
+
+def endpoint_to_dict(endpoint):
+    return dict(
+        deployment_status = endpoint.deployment_status,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: unexpected spaces around keyword / parameter equals

Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
        deployment_status=endpoint.deployment_status,
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+        enabled_state = endpoint.enabled_state,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: The same below
```suggestion
        enabled_state=endpoint.enabled_state,
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+    def __init__(self):
+        self.module_arg_spec = dict(
+            auto_generated_domain_name_label_scope=dict(
+                type='str',
+                default = 'TenantReuse',
+                choices=["TenantReuse", "SubscriptionReuse", "ResourceGroupReuse", "NoReuse"]
+            ),
+            enabled_state=dict(
+                type='str',
+                default = 'Enabled',
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
                default='Enabled',
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+                type='str',
+                required=True
+            ),
+            state=dict(
+                type='str',
+                default='present',
+                choices=['present', 'absent']
+            )
+        )
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion

```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+        self.profile_name = None
+        self.resource_group = None
+        self.state = None
+
+        self.endpoint_client = None
+
+        self.results = dict(changed=False)
+
+        super(AzureRMEndpoint, self).__init__(derived_arg_spec=self.module_arg_spec,
+                                                supports_check_mode=True,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: 177， 178 continuation line over-indented for visual indent

File: plugins/modules/azure_rm_afdendpoint.py [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+
+                if not self.check_mode:
+                    new_response = self.create_endpoint()
+                    self.results['id'] = new_response['id']
+                    self.results['host_name'] = new_response['host_name']
+                    self.log("AFD Endpoint creation done")
+
+                self.results['changed'] = True
+                return self.results
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion

```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+                    self.results['id'] = new_response['id']
+                    self.results['host_name'] = new_response['host_name']
+                    self.log("AFD Endpoint creation done")
+
+                self.results['changed'] = True
+                return self.results
+            
+            else:
+                self.log('Results : {0}'.format(response))
+                self.results['id'] = response['id']
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
                self.results['id'] = response['id']
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+                    self.log("AFD Endpoint creation done")
+
+                self.results['changed'] = True
+                return self.results
+            
+            else:
+                self.log('Results : {0}'.format(response))
+                self.results['id'] = response['id']                
+                self.results['host_name'] = response['host_name']
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
                
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+                
+                update_tags, self.tags = self.update_tags(response['tags'])
+
+                if update_tags:
+                    to_be_updated = True
+
+                if response['provisioning_state'] == "Succeeded":
+                    if response['enabled_state'] != self.enabled_state:
+                        to_be_updated = True
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion

```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+                        self.log("AFD Endpoint locations changes are not idempotent, please delete and recreate the Endpoint.")
+
+                    if to_be_updated:
+                        self.log("Need to update the AFD Endpoint")
+                        self.results['changed'] = True
+
+                        if not self.check_mode:
+                            result = self.update_endpoint()
+                            self.results['host_name'] = result['host_name']
+                            self.log("AFD Endpoint update done")
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
                            self.log("AFD Endpoint update done")
```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+
+        :return: deserialized AFD Endpoint instance state dictionary
+        '''
+        self.log("Updating the AFD Endpoint instance {0}".format(self.name))
+
+        parameters = AFDEndpointUpdateParameters(
+            tags=self.tags,
+            enabled_state=self.enabled_state
+        )
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion

```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+            self.log("AFD Endpoint : {0} found".format(response.name))
+            return endpoint_to_dict(response)
+        except Exception as err:
+            self.log('Did not find the AFD Endpoint.')
+            return False
+
+    def get_endpoint_client(self):
+        if not self.endpoint_client:
+            self.endpoint_client = self.get_mgmt_svc_client(CdnManagementClient,
+                                                       base_url=self._cloud_environment.endpoints.resource_manager,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]: 341， 342 continuation line under-indented for visual indent

File: plugins/modules/azure_rm_afdendpoint.py [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+                                                       api_version='2023-05-01')
+        return self.endpoint_client
+
+
+def main():
+    """Main execution"""
+    AzureRMEndpoint()
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion


```

File: plugins/modules/azure_rm_afdendpoint.py [hunk 25]
```diff
@@ -0,0 +1,351 @@
+#!/usr/bin/python
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#
+# Python SDK Reference: https://learn.microsoft.com/en-us/python/api/azure-mgmt-cdn/azure.mgmt.cdn.operations.afdendpointsoperations?view=azure-python
+#
+# TODO: Name check the URL
+#
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_afdendpoint.py]:
```suggestion
#
```

**Comment by Fred-sun** [comment]
@jartoo Please fix the sanity error!  Thank you!

File: plugins/modules/azure_rm_afdendpoint.py (hunk 8 above)
Comment (jartoo) [reply plugins/modules/azure_rm_afdendpoint.py:16]: Was not sure what version to assign it...putting in 2.5.0 unless I get further guidance.  Thanks!

File: plugins/modules/azure_rm_afdendpoint.py (hunk 18 above)
Comment (jartoo) [reply plugins/modules/azure_rm_afdendpoint.py]: Not sure I understand this one, as there is no response in this section of code, so setting self.results['id'] = response['id'] would be blank.  I set it to new_response['id'] as the return result from the create_endpoint() function.  I could honestly be missing something though..  :)

Comment (jartoo) [reply plugins/modules/azure_rm_afdendpoint.py]: Or are you referring to having a value when in check mode?

**Comment by jartoo** [comment]
Sorry, I am just now sorting out the sanity tests...I see where you are getting most of this from now!  Will work these issues and cycle back with corrections.

**Comment by Fred-sun** [comment]
@jartoo In addition, we can prioritize the two space bar problems in #1588, and we can prioritize the merger after solving them. Because the following several PR are dependent on its properties! Thank you

**Comment by Fred-sun** [comment]
@xuzhang3 This PR relies on #1588, Please merge #1588 at first！Thanks!

**Review by xuzhang3 — APPROVED** [review]

**Pull request merged**